service/http.proxy/src/main/java/org/kaazing/gateway/service/http/proxy/HttpProxyServiceHandler.java

```diff
@@ -94,10 +95,31 @@ public void initializeSession(IoSession session, ConnectFuture future) {
             HttpConnectSession connectSession = (HttpConnectSession) session;
             connectSession.setVersion(acceptSession.getVersion());
             connectSession.setMethod(acceptSession.getMethod());
-            connectSession.setRequestURI(acceptSession.getRequestURI());
+            if (!validateRequestPath()) {
+                acceptSession.setStatus(CLIENT_NOT_FOUND);
+                acceptSession.close(false);
+                return;
+            }
+            String acceptPathInfo = acceptSession.getPathInfo().toString();
+            if (acceptPathInfo.indexOf("/") == 0) {
+                acceptPathInfo = acceptPathInfo.substring(1);
+            }
+            URI connectRequestURI = connectSession.getRequestURI().resolve(acceptPathInfo);
+            connectSession.setRequestURI(connectRequestURI);
             processRequestHeaders(acceptSession, connectSession);
         }
 
+        private boolean validateRequestPath() {
+            URI requestURI = acceptSession.getRequestURI();
+            String acceptPath = acceptSession.getServicePath().getPath();
+            String requestPath = requestURI.normalize().getPath();
+            if (requestPath.indexOf("/\\.\\./") == 0 || requestPath.indexOf(acceptPath) != 0) {
```

Comment:
I think here the tcp connection is already established to origin server, i.e connect session is established. Then we are doing the validation of accept path and deciding not to send http request. I think it is better to do the validation in the sessionOpened(), that way gateway doesn't connect to origin server for invalid requests.